File: LeetCode/P347_TopKFrequentElements.java
/*
 * LeetCode 347 - Top K Frequent Elements
 * https://leetcode.com/problems/top-k-frequent-elements/
 *
 * Approach: Bucket Sort by frequency
 * Time  : O(n)
 * Space : O(n)
 */

import java.util.*;

public class P347_TopKFrequentElements {

    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.merge(num, 1, Integer::sum);
        }

        // bucket[i] holds all numbers that appear exactly i times
        List<Integer>[] bucket = new List[nums.length + 1];
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            int freq = entry.getValue();
            if (bucket[freq] == null) {
                bucket[freq] = new ArrayList<>();
            }
            bucket[freq].add(entry.getKey());
        }

        List<Integer> result = new ArrayList<>();
        for (int i = bucket.length - 1; i >= 0 && result.size() < k; i--) {
            if (bucket[i] != null) {
                result.addAll(bucket[i]);
            }
        }

        return result.stream().mapToInt(Integer::intValue).limit(k).toArray();
    }

    // ---------- Simple manual test ----------
    public static void main(String[] args) {
        P347_TopKFrequentElements sol = new P347_TopKFrequentElements();

        int[] nums1 = {1, 1, 1, 2, 2, 3};
        System.out.println(Arrays.toString(sol.topKFrequent(nums1, 2))); // [1, 2]

        int[] nums2 = {1};
        System.out.println(Arrays.toString(sol.topKFrequent(nums2, 1))); // [1]
    }
}

File: LeetCode/P347_TopKFrequentElements.cpp
#include <bits/stdc++.h>
using namespace std;

class P347_TopKFrequentElements {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freqMap;
        for (int num : nums) {
            freqMap[num]++;
        }

        vector<vector<int>> bucket(nums.size() + 1);
        for (const auto& entry : freqMap) {
            int freq = entry.second;
            bucket[freq].push_back(entry.first);
        }

        vector<int> result;
        for (int i = (int) bucket.size() - 1; i >= 0 && (int) result.size() < k; i--) {
            for (int value : bucket[i]) {
                result.push_back(value);
            }
        }

        if ((int) result.size() > k) {
            result.resize(k);
        }
        return result;
    }
};

string toString(const vector<int>& arr) {
    ostringstream sb;
    sb << "[";
    for (size_t i = 0; i < arr.size(); i++) {
        if (i > 0) {
            sb << ", ";
        }
        sb << arr[i];
    }
    sb << "]";
    return sb.str();
}

int main() {
    P347_TopKFrequentElements sol;

    vector<int> nums1 = {1, 1, 1, 2, 2, 3};
    cout << toString(sol.topKFrequent(nums1, 2)) << "\n";

    vector<int> nums2 = {1};
    cout << toString(sol.topKFrequent(nums2, 1)) << "\n";

    return 0;
}
